examples/end-to-end-examples/support_chatbot/chatbot.py

```diff
@@ -0,0 +1,546 @@
+                continuation = data.get('continuation')
+                response = chain(question)
+                answer = self.clean_response(response['result'].lstrip('\n'))
+                time.sleep(0.5)
+                save += f'Question:\n{question}\nAnswer:\n{continuation}\nResponse:\n{answer}\n\n'
+        return save
+    
+    def sub_query_chat(self,
+                       query: str)-> str:
+        if not self.intent_chain:
+            save_k = self.k
+            self.k = 5
+            self.intent_chain = self.create_chain(SUBQUERY_INTENT_TEMPLATE)
+            self.k = save_k
+        intent_response = self.intent_chain(query)
+        intent_answer = self.clean_response(intent_response['result'].lstrip('\n'))
+        
+        SUBQUERY_SUBQA_TEMPLATE = PARTIAL_SUBQA_TEMPLATE.format(intent_answer)
+        subQA_chain = self.create_chain(SUBQUERY_SUBQA_TEMPLATE)
+        subQA_response = subQA_chain(query)
+        subQA_answer = self.clean_response(subQA_response['result'].lstrip('\n'))
+
+        all_sub_QA = subQA_answer.split('\n')
+        sub_QA_injection = ''
+        # Don't create a new chain on every query
+        if not self.subchain:
+            self.subchain = self.create_chain(EVAL_30B_TEMPLATE)
+        for sub_QA in all_sub_QA:
+            response = self.subchain(sub_QA[1:])
+            if response['source_documents']:
+                answer = self.clean_response(response['result'].lstrip('\n'))
+                sub_QA_injection += f'Question: {sub_QA[1:]} \nAnswer: {answer}\n'
+
+        SUBQUERY_COMBINE_TEMPLATE = PARTIAL_COMBINE_TEMPLATE.format(sub_QA_injection)
+        combine_chain = self.create_chain(SUBQUERY_COMBINE_TEMPLATE)
+        combine_response = combine_chain(query)
+        combine_answer = self.clean_response(combine_response['result'].lstrip('\n'))
+        sources = ''
+        slack_deduplicate = True
+        for d in combine_response['source_documents']:
+            if d.metadata["score"] < 0.7:
+                if 'message_from_slack' == sources[:18] and slack_deduplicate:
+                    sources = sources + 'slack_data\n'
+                    slack_deduplicate = False
+                else:
+                    sources = sources + f'{d.metadata["file_name"].replace("{slash}", "/")}\n'
+        if not sources:
+            return f'Answer: \n{combine_answer}\n\nIntent: \n{intent_answer}\n\nRelated Sub-questions: \n{sub_QA_injection}'
+        else:
+            return f'Answer: \n{combine_answer}\n\nIntent: \n{intent_answer}\n\nRelated Sub-questions: \n{sub_QA_injection}\nSources: \n{sources}'
+          
+    def chat(self, 
+             query: str) -> str:
+        """Chat with the chatbot given a query
+        
+        Args:
+            query (str): The query to ask the chatbot
+        """
+    
+        if query == "!eval_7b":
+            self.set_eval_state(endpoint_url='https://chatbot-7b-finetuned-rxyfc5.inf.hosted-on.mosaicml.hosting/predict')
```

Comment:
probably need to accept the endpoint urls as an arg instead of hardcoding to an endpoint that won't exist for a user.